### ISSUE TITLE
How do you figure out parent and child relations between pages in gatsbyjs?

### ISSUE THREAD

**Issue by Megajin** [gatsbyjs/gatsby#26752]
original question was asked at Stackoverflow: https://stackoverflow.com/questions/63674700/how-do-you-figure-out-parent-and-child-relations-between-pages-in-gatsbyjs

## Summary

**Is there any logical and easy way to tell gatsby that a given page in a sub folder is a child page?**

I know that if you put pages in sub folders under `/src/pages/subpages`, gatsby will produce a route under the given folder like:

```
/src/pages/page-01.js
/src/pages/page-02/index.js
/src/pages/page-02/child-page-of-02.js
```
will give you:
```
http://localhost:8000/page-01
http://localhost:8000/page-02
http://localhost:8000/page-02/child-page-of-02
```

Now, I've tried to easily fill a sidebar with page links and wanted to indent the child pages.
To do so I've used the query:

```
query MyQuery {
  allSitePage {
    nodes {
      id
      children {
        id
      }
    }
  }
}

```

The results were:

```
{
  "data": {
    "allSitePage": {
      "nodes": [
        {
          "id": "SitePage /page-01",
          "children": []
        },
        {
          "id": "SitePage /dev-404-page/",
          "children": []
        },
        {
          "id": "SitePage /",
          "children": []
        },
        {
          "id": "SitePage /page-02/",
          "children": []
        },
        {
          "id": "SitePage /page-02/child-page-of-02/",
          "children": []
        }
      ]
    }
  },
  "extensions": {}
}
```
As you can see, the children array is empty. Which is not what I was expecting.
I am expecting this output:

```
        {
          "id": "SitePage /page-02/",
          "children": ["SitePage /page-02/child-page-of-02/"]
        },
        {
          "id": "SitePage /page-02/child-page-of-02/",
          "children": []
        }
```

Has anyone got this working? If yes, how?
I did already lookup in the docs but could not find anything useful...

Nothing in: [Page Creation][1] and [Node Creation][2]

  [1]: https://www.gatsbyjs.com/docs/page-creation/
  [2]: https://www.gatsbyjs.com/docs/node-creation/

### Environment
  ```
  System:
    OS: Windows 10 10.0.18363
    CPU: (8) x64 Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
  Binaries:
    Node: 12.3.1 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.4 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.9.0 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 2.7.16 - /c/Python27/python
  Browsers:
    Chrome: 85.0.4183.83
    Edge: Spartan (44.18362.449.0)
  npmPackages:
    gatsby: ^2.24.52 => 2.24.52 
    gatsby-image: ^2.4.16 => 2.4.16 
    gatsby-plugin-advanced-sitemap: ^1.5.5 => 1.5.5 
    gatsby-plugin-catch-links: ^2.3.11 => 2.3.11 
    gatsby-plugin-manifest: ^2.4.26 => 2.4.26 
    gatsby-plugin-material-ui: ^2.1.10 => 2.1.10 
    gatsby-plugin-react-helmet: ^3.3.10 => 3.3.10 
    gatsby-plugin-sharp: ^2.6.30 => 2.6.30 
    gatsby-plugin-web-font-loader: ^1.0.4 => 1.0.4 
    gatsby-source-filesystem: ^2.3.27 => 2.3.27 
    gatsby-transformer-remark: ^2.8.31 => 2.8.31 
    gatsby-transformer-sharp: ^2.5.13 => 2.5.13 
  npmGlobalPackages:
    gatsby-cli: 2.12.89

```

### File contents

`gatsby-config.js`:
```javascript
module.exports = {
  siteMetadata: {
    title: `my-test-page.com`,
    description: `word, word, word`,
    author: `https://github.com/Megajin,`,
    siteUrl: `https://my-test-page.com`
  },
  plugins: [
    `gatsby-plugin-catch-links`,
    `gatsby-plugin-react-helmet`,
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `images`,
        path: `${__dirname}/src/images`
      }
    },
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `markdown`,
        path: `${__dirname}/src/markdown/`
      }
    },
    {
      resolve: 'gatsby-plugin-web-font-loader',
      options: { google: { families: ['Goblin One'] } }
    },
    {
      resolve: `gatsby-plugin-sharp`,
      options: {
        useMozJpeg: false,
        stripMetadata: true
        // defaultQuality: 100
      }
    },
    `gatsby-transformer-sharp`,
    {
      resolve: `gatsby-plugin-manifest`,
      options: {
        name: `my-test-page.com`,
        short_name: `my-test-page`,
        start_url: `/`,
        lang: `de`,
        display: `minimal-ui`,
        icon: `./static/Icon.png`
      }
    },
    {
      resolve: 'gatsby-transformer-remark',
      gfm: true,
      options: { plugins: [] }
    },
    {
      resolve: `gatsby-plugin-material-ui`,
      options: { stylesProvider: { injectFirst: true } }
    },
    `gatsby-plugin-advanced-sitemap`
    // `gatsby-plugin-offline`
    // this (optional) plugin enables Progressive Web App + Offline functionality
    // To learn more, visit: https://gatsby.dev/offline
  ]
};
``` 
`package.json`: 
```javascript
{
  "name": "my-test-page",
  "private": true,
  "description": "my-test-page.com",
  "version": "1.0.0",
  "license": "MIT",
  "scripts": {
    "build": "gatsby build",
    "develop": "gatsby develop",
    "format": "prettier --write \"**/*.{js,jsx,ts,tsx,json,md}\"",
    "start": "npm run develop",
    "serve": "gatsby serve",
    "clean": "gatsby clean",
    "test": "echo \"Write tests! -> https://gatsby.dev/unit-testing\" && exit 1"
  },
  "dependencies": {
    "@material-ui/core": "^4.11.0",
    "@material-ui/icons": "^4.9.1",
    "@material-ui/styles": "^4.10.0",
    "@mdi/js": "^5.5.55",
    "@mdi/react": "^1.4.0",
    "@mdi/svg": "^5.5.55",
    "gatsby": "^2.24.52",
    "gatsby-image": "^2.4.16",
    "gatsby-plugin-advanced-sitemap": "^1.5.5",
    "gatsby-plugin-catch-links": "^2.3.11",
    "gatsby-plugin-manifest": "^2.4.26",
    "gatsby-plugin-material-ui": "^2.1.10",
    "gatsby-plugin-react-helmet": "^3.3.10",
    "gatsby-plugin-sharp": "^2.6.30",
    "gatsby-plugin-web-font-loader": "^1.0.4",
    "gatsby-source-filesystem": "^2.3.27",
    "gatsby-transformer-remark": "^2.8.31",
    "gatsby-transformer-sharp": "^2.5.13",
    "notistack": "^1.0.0",
    "prop-types": "^15.7.2",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "react-helmet": "^6.1.0",
    "react-redux": "^7.2.1",
    "redux": "^4.0.5",
    "redux-thunk": "^2.3.0",
    "uuid": "^8.3.0"
  },
  "devDependencies": {
    "eslint": "^7.7.0",
    "eslint-config-airbnb": "^18.2.0",
    "eslint-plugin-import": "^2.22.0",
    "eslint-plugin-jsx-a11y": "^6.3.1",
    "eslint-plugin-react": "^7.20.6",
    "eslint-plugin-react-hooks": "^4.1.0",
    "prettier": "2.1.0"
  }
}
```
`gatsby-node.js`: 
```javascript
const path = require('path');

exports.createPages = async ({ actions, graphql, reporter }) => {
  const { createPage } = actions;
  const singlePageTemplate = path.resolve(`src/templates/single-page.js`);
  const result = await graphql(`
    {
      allMarkdownRemark(sort: { order: ASC, fields: [frontmatter___order] }, limit: 1000) {
        edges {
          node {
            frontmatter {
              path
            }
          }
        }
      }
    }
  `);
  if (result.errors) {
    reporter.panicOnBuild(`Error while running GraphQL query.`);
    return;
  }
  result.data.allMarkdownRemark.edges.forEach(({ node }) => {
    createPage({
      path: node.frontmatter.path,
      component: singlePageTemplate,
      context: {} // additional data can be passed via context
    });
  });
};
```

`gatsby-browser.js & gatsby-ssr.js`:
```javascript
import React from 'react';
import PropTypes from 'prop-types';
import CssBaseline from '@material-ui/core/CssBaseline';

import { Provider } from 'react-redux';
import { SnackbarProvider } from 'notistack';
import { MuiThemeProvider } from '@material-ui/core/styles';
import { configureStore } from './src/redux/configure-store';
import { theme } from './src/theme/mui-theme';
/**
 * Creates Store for Redux Pattern.
 * Only use one store for the whole app!
 * defines state values.
 * define store with all reducers needed for the app.
 * so basically every single one under ./reducers/*.
 * @type {Store}
 */
const store = configureStore();

export const wrapRootElement = ({ element }) => (
  <Provider store={store}>
    <MuiThemeProvider theme={theme}>
      <CssBaseline />
      <SnackbarProvider
        maxSnack={3}
        preventDuplicate
        anchorOrigin={{
          vertical: 'top',
          horizontal: 'center'
        }}
        autoHideDuration={3500}
      >
        {element}
      </SnackbarProvider>
    </MuiThemeProvider>
  </Provider>
);

wrapRootElement.propTypes = { element: PropTypes.node.isRequired };
```

**Comment by wardpeet** [comment]
Pages our entities on their own, there is no linking between them from a Gatsby data point of view. You can accomplish what you want through https://www.gatsbyjs.com/docs/schema-customization/ and create your own PageNode structure.

**Comment by vladar** [comment]
There are several related issues on the same subject: #4583 #20331

I think this is a good feature for a plugin! There is an outdated plugin that does this: https://github.com/anotherjames/gatsby-plugin-page-tree/blob/master/src/gatsby-node.ts

But it uses old `setFieldsOnGraphQLNodeType` API, if you're up to it, you should be using a newer [schema customization API](https://www.gatsbyjs.com/docs/schema-customization/) as advised by @wardpeet

**Comment by vladar** [comment]
We're marking this issue as answered and closing it for now but please feel free to comment if you would like to continue this discussion. Thanks for using Gatsby 💜

**Comment by Megajin** [comment]
Thanks guys, I will look into the provided informations 👍 .

**Comment by Megajin** [comment]
For anyone coming by, I've made a plugin for this purpose: https://github.com/dream-bit-de/gatsby-plugin-better-page-tree
